From 7d70a53108ac1ab5cdd817b3bb3b0426b96a0c9c Mon Sep 17 00:00:00 2001
From: Daniel Borkmann <dborkman@redhat.com>
Date: Fri, 20 Jun 2014 10:37:21 -0400
Subject: [net] sctp: check proc_dointvec result in proc_sctp_do_auth

Message-id: <56a684ddc130a2caa82b7bab2ce385b1bbf4be01.1403260378.git.dborkman@redhat.com>
Patchwork-id: 84294
O-Subject: [RHEL7 PATCH net 7/5] net: sctp: check proc_dointvec result in proc_sctp_do_auth
Bugzilla: 1110290
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110290
Upstream Status: 24599e61b7552673dd85971cf5a35369cd8c119e (net)
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7604375
Tested: verified on beaker

Upstream Commit:
commit 24599e61b7552673dd85971cf5a35369cd8c119e
Author: Daniel Borkmann <dborkman@redhat.com>
Date:   Wed Jun 18 23:46:31 2014 +0200

    net: sctp: check proc_dointvec result in proc_sctp_do_auth

    When writing to the sysctl field net.sctp.auth_enable, it can well
    be that the user buffer we handed over to proc_dointvec() via
    proc_sctp_do_auth() handler contains something other than integers.

    In that case, we would set an uninitialized 4-byte value from the
    stack to net->sctp.auth_enable that can be leaked back when reading
    the sysctl variable, and it can unintentionally turn auth_enable
    on/off based on the stack content since auth_enable is interpreted
    as a boolean.

    Fix it up by making sure proc_dointvec() returned sucessfully.

    Fixes: b14878ccb7fa ("net: sctp: cache auth_enable per endpoint")
    Reported-by: Florian Westphal <fwestpha@redhat.com>
    Signed-off-by: Daniel Borkmann <dborkman@redhat.com>
    Acked-by: Neil Horman <nhorman@tuxdriver.com>
    Acked-by: Vlad Yasevich <vyasevich@gmail.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Daniel Borkmann <dborkman@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/sctp/sysctl.c b/net/sctp/sysctl.c
index 3a72606..e81312b 100644
--- a/net/sctp/sysctl.c
+++ b/net/sctp/sysctl.c
@@ -454,8 +454,7 @@ static int proc_sctp_do_auth(struct ctl_table *ctl, int write,
   tbl.data = &net->sctp.auth_enable;
 
  ret = proc_dointvec(&tbl, write, buffer, lenp, ppos);
-
- if (write) {
+ if (write && ret == 0) {
   struct sock *sk = net->sctp.ctl_sock;
 
   net->sctp.auth_enable = new_value;
-- 
1.7.1